### PR TITLE
Dsl/moist/radiation coupling - RadCouple Port in pyMoist

### PR DESCRIPTION
**Pull request by mattweil11**
Ticket [NDSL] Radiation Coupling #937 
do_qa not implemented in radiation_coupling.py
Waiting on up to date testing data
Tests fail for CLDREFFL (5.55%) and CLDREFFI (5.08%)
Max Error set at 1e-9

### CODE REVIEW

**Comment by FlorianDeconinck** [comment]
0 diff explainer: this is related to the port of the Moist physics to the NDSL stack, it's execution and compilation is deactivated by default

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/constants.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+MAPL_SRFPRS = 98470  # Pa
+MAPL_KARMAN = 0.40  # --
+MAPL_USMIN = 1.00  # m/s
+MAPL_RHO_SEAWATER = 1026.0  # sea water density [kg/m^3]
+MAPL_RHO_SEAICE = 917.0  # sea ice density [kg/m^3]
+MAPL_RHO_SNOW = 330.0  # snow density [kg/m^3]
+MAPL_CELSIUS_TO_KELVIN = 273.15  # K
+
+# Define constants for 
+alhlbcp = MAPL_ALHL / MAPL_CP
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/constants.py:77]: ALL CAPS for global constants

Look through the code.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py [hunk 2]
```diff
@@ -1,32 +1,287 @@
...
 
+#Calculate air density [kg/m^3]
+@gtscript.function
+def RHO(PL, TE):
+    RHO = (100.0 * PL) / (radconstants.MAPL_RGAS * TE)
+    return RHO
 
-def _fix_up_clouds_stencil():
-    pass
+#Calculate ice cloud effective radius [m]
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:14]: Rename to `cloud_effective_radius_ice` and put a docstring to document the match to LDRADIUS4_X.

See docstrings standard: https://peps.python.org/pep-0257/

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py [hunk 3]
```diff
@@ -1,32 +1,287 @@
...
+@gtscript.function
+def LDRADIUS4_ICE(
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:16]: See above.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py [hunk 4]
```diff
@@ -1,32 +1,287 @@
...
+        ZFSR = 1.2351 + 0.0105 * TC
+        AA = 45.8966 * (WC**0.2214)
+        BB = 0.79570 * (WC**0.2535)
+        RADIUS = ZFSR * (AA + BB * (TE - 83.15))
+        RADIUS = min(150.0e-6, max(5.0e-6, 1.0e-6 * RADIUS * 0.64952))
+    return RADIUS
 
+#Calculate liquid cloud effective radius [m]
+@gtscript.function
+def LDRADIUS4_LIQUID(
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:47]: See above.

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py [hunk 5]
```diff
@@ -1,32 +1,287 @@
...
+                        MAX_RI = MAX_RI,
+                        )
+        '''
+        From GEOS_GFDL_1M_InterfaceMod.F90 Line 866 not implemented
+        Implement QSAT0 and QSAT3 logic if diagnostics are needed. 
+        QSAT0 and QSAT3 are found in GEOS/src/Shared/@GMAO_Shared/GEOS_Shared/GEOS_Utilities.F90 
+        
+        Variables: RHX - FloatField
+        '''
         if self.do_qa:
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:285]: Move the `do_qa` handling in `__init__` and live a one liner comment here

**Review by FlorianDeconinck — CHANGES_REQUESTED** [review]

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py [hunk 6]
```diff
@@ -1,32 +1,407 @@
-from ndsl import QuantityFactory, StencilFactory
 from ndsl.constants import X_DIM, Y_DIM, Z_DIM
+from ndsl.dsl.typing import FloatField, Int, Float 
+from ndsl import QuantityFactory, StencilFactory
+import gt4py.cartesian.gtscript as gtscript
+from gt4py.cartesian.gtscript import computation, interval, PARALLEL, log10
+import pyMoist.radiation_coupling_constants as radconstants
+
+@gtscript.function
+def RHO(
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:9]: Rename air_density

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py [hunk 7]
```diff
@@ -1,32 +1,407 @@
...
         self.do_qa = do_qa
-
-    def __call__(self) -> None:
-        self._fix_up_clouds()
-        self._radcouple()
+        #GEOS_GFDL_1M_InterfaceMod.F90:866 not implemented. Implement QSAT0 and QSAT3 logic if diagnostics are needed, found in GEOS/src/Shared/@GMAO_Shared/GEOS_Shared/GEOS_Utilities.F90.
         if self.do_qa:
-            pass
+            #RHX = Q/GEOS_QSAT( T, PLmb)
+            raise NotImplementedError("Module procedures QSAT0 and QSAT3 that are needed for GEOS_QSAT are not implemented")
```
Comment (FlorianDeconinck) [inline GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:302]: "[Radiation Coupling] Diagnostic (do_qa) not implemented. (GEOS_QSAT missing)"

**Review by FlorianDeconinck — APPROVED** [review]
Couple of minor things, then we can merge

File: GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py (hunk 6 above)
Comment (FlorianDeconinck) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:9]: Unless `rho` is a common name for air_density I guess

Comment (mattweil11) [reply GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSmoist_GridComp/pyMoist/pyMoist/radiation_coupling.py:9]: rho is common name for density in general. Will change to air_density to specify

**Pull request merged**